### PR TITLE
enhance searchcv tool and upgrade of sklearn to v0.20.0

### PR DESCRIPTION
**Pull request by qiagu**
1) make explicit cv splitter
2) support more estimator searches, like the `base_estimator` of `AdaBoostRegressor`
3) support preprocessor swap/search
4) make preprocessor hot number/keyword

### CODE REVIEW

File: tools/sklearn/main_macros.xml [hunk 1]
```diff
@@ -244,8 +244,8 @@
     <param argument="shuffle" type="boolean" optional="true" truevalue="booltrue" falsevalue="boolfalse" checked="@CHECKED@" label="@LABEL@" help="@HELP_TEXT@"/>
   </xml>
 
-  <xml name="random_state" token_default_value="" token_help_text="Integer number. The seed of the pseudo random number generator to use when shuffling the data. A fixed seed allows reproducible results.">
-    <param argument="random_state" type="integer" optional="true" value="@DEFAULT_VALUE@" label="Random seed number" help="@HELP_TEXT@"/>
+  <xml name="random_state" token_default_value="" token_help_text="Integer number. The seed of the pseudo random number generator to use when shuffling the data. A fixed seed allows reproducible results. default=None.">
+    <param argument="random_state" type="integer" optional="true" value="@DEFAULT_VALUE@" label="random_state" help="@HELP_TEXT@"/>
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: Why changing the label here ... a label with underscore looks wired.

File: tools/sklearn/main_macros.xml [hunk 2]
```diff
@@ -825,10 +825,114 @@
...
+      </when>
+      <when value="LeavePGroupsOut">
+        <param argument="n_groups" type="integer" value="" label="n_groups" help="Number of groups (p) to leave out in the test split." />
+        <expand macro="cv_groups"/>
+      </when>
+    </conditional>
+  </xml>
+
+  <xml name="cv_n_splits" token_value="3" token_help="Number of folds. Muast be at least 2.">
+    <param argument="n_splits" type="integer" value="@VALUE@" label="n_splits" help="@HELP@"/>
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: When it needs to be at least 2, then remove this text and use `min="2"`

File: tools/sklearn/main_macros.xml [hunk 3]
```diff
@@ -825,10 +825,114 @@
...
+  </xml>
+
+  <xml name="cv_shuffle">
+    <param argument="shuffle" type="boolean" optional="true" truevalue="booltrue" falsevalue="boolfalse" checked="false" label="shuffle" />
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: The label can be enhanced a little bit. What is to shuffle etc ..

File: tools/sklearn/main_macros.xml [hunk 4]
```diff
@@ -825,10 +825,114 @@
...
+  </xml>
+
+  <xml name="cv_test_size" token_value="0.2">
+    <param argument="test_size" type="float" value="@VALUE@" label="test_size" help="0.0-1.0, proportion of the dataset to include in the test split; >1, integer only, the absolute number of test samples "/>
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: remove `0.0-1.0` and use `min="0.0" max="1.0"`

File: tools/sklearn/main_macros.xml [hunk 5]
```diff
@@ -825,10 +825,114 @@
...
+  </xml>
+
+  <xml name="cv_groups" >
+    <param argument="groups" type="text" value="" area="true" label="groups" help="Group lables in a list. e.g., [1, 1, 2, 2, 3, 3, 3]">
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: indentation

Comment (bgruening) [reply tools/sklearn/main_macros.xml]: Is the bracket really needed, we can add it later isn't it?

File: tools/sklearn/model_validation.xml [hunk 6]
```diff
@@ -67,7 +68,16 @@ y = read_columns(
 y=y.ravel()
 
 options = params["model_validation_functions"]["options"]
-options['cv'] = get_cv( options['cv'] )
+splitter, groups = get_cv(options.pop('cv_selector'))
+if groups is not None:
+    groups = groups.strip()
+    if groups == "":
+        groups = None
```
Comment (bgruening) [inline tools/sklearn/model_validation.xml]: How does an input look like if this is true and why do we pass then None to the splitter ... is adding `  ` a actual usecase?

File: tools/sklearn/main_macros.xml (hunk 1 above)
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: wow, I think you're right. Explicit labels sound better.

File: tools/sklearn/main_macros.xml (hunk 2 above)
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: Good idea. I think we still can keep the `min="2"` in the help.

File: tools/sklearn/main_macros.xml (hunk 4 above)
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: We can add `min="0.0"`. `test_size` takes both float, 0.0-1.0, and integer, greater than 1.0. When integer, represent the number of instances in tests.

File: tools/sklearn/main_macros.xml (hunk 5 above)
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: indentation?

I have an idea. Using string.split to generate the list. I prefer to keep the `[` and `]`, but remove them ( will be __fb__ and __rb__) before calling split. What do you think?

File: tools/sklearn/model_validation.xml (hunk 6 above)
Comment (qiagu) [reply tools/sklearn/model_validation.xml]: It's consistent to cast empty string to `None` and may be helpful in debugging.

**Comment by qiagu** [comment]
I'll make changes later today. Thanks

File: tools/sklearn/main_macros.xml (hunk 3 above)
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: Recently, I began to like short labels, especially for those key words existing in scikit-learn, maybe just because I am a heavy user. I also thought simple words made the UI clean. But I'll value your opinion and other people's as well.

File: tools/sklearn/model_validation.xml (hunk 6 above)
Comment (bgruening) [reply tools/sklearn/model_validation.xml]: Ok, if you think this is more readable. Should we use this as a macro?

Comment (qiagu) [reply tools/sklearn/model_validation.xml]: I'll look into this later, thanks.

File: tools/sklearn/search_model_validation.xml [hunk 7]
```diff
@@ -551,7 +667,118 @@ The evaluation of settings supports operations in Math, list comprehension, nump
 
 - max_depth: scipy_stats_randin(1, 11)
 
-- estimator: [ensemble_ExtraTreesClassifier(n_estimators=100, random_state=324089)]
+**Estimator search/eval (additional '-'):**
+
+- base_estimator-: [sklearn_tree.DecisionTreeRegressor(), sklearn_tree.ExtraTreeRegressor()]
```
Comment (bgruening) [inline tools/sklearn/search_model_validation.xml]: I find this `-:` syntax very confusing. Can you explain this in more detail, please.

File: tools/sklearn/search_model_validation.xml [hunk 8]
```diff
@@ -551,7 +667,118 @@ The evaluation of settings supports operations in Math, list comprehension, nump
...
+
+'decomp_all': All decomposition estimators, i.e., 15-25
+
+'k_appr_all': All kernel_approximation estimators, i.e. 26-29
+
+'reb_all': All skrebate estimators, i.e. 31-35
+
+'imb_all': All imbalanced-learn sampling methods, i.e. 36-54. CAUTION: Mix of imblearn and other preprocessors may not work.
+
+'all': All, 0-36
```
Comment (bgruening) [inline tools/sklearn/search_model_validation.xml]: None, is not explained.

File: tools/sklearn/search_model_validation.xml [hunk 9]
```diff
@@ -551,7 +667,118 @@ The evaluation of settings supports operations in Math, list comprehension, nump
...
+
+Support mix (except imblearn methods), e.g.:
+
+- : [None, 'sk_prep_all', 22, 'k_appr_all', sklearn_feature_selection.SelectKBest(k=50)]
```
Comment (bgruening) [inline tools/sklearn/search_model_validation.xml]: the `- :` is part of the parameter or not? Why do we not offer these 35 options in a select field? Or in a repeat?

File: tools/sklearn/utils.py [hunk 10]
```diff
@@ -222,11 +254,112 @@ def get_search_params(params_builder):
...
+                            imblearn.over_sampling.SMOTE(random_state=0, n_jobs=N_JOBS),
+                            imblearn.over_sampling.SVMSMOTE(random_state=0, n_jobs=N_JOBS),
+                            imblearn.over_sampling.BorderlineSMOTE(random_state=0, n_jobs=N_JOBS),
+                            imblearn.over_sampling.SMOTENC(categorical_features=[], random_state=0, n_jobs=N_JOBS),
+                            imblearn.combine.SMOTEENN(random_state=0), imblearn.combine.SMOTETomek(random_state=0)]
+            i = 0
+            while i < len(ev):
+                obj = ev[i]
+                if obj is None:
+                    i = i + 1
```
Comment (bgruening) [inline tools/sklearn/utils.py]: i += 1

File: tools/sklearn/utils.py [hunk 11]
```diff
@@ -222,11 +254,112 @@ def get_search_params(params_builder):
...
+                    if 'n_jobs' in obj.get_params():
+                        obj.set_params( n_jobs=N_JOBS )
+                if param_type == "final_estimator_p":
+                    search_params["estimator__" + param_name[:-1]] = ev
+                else:
+                    search_params["preprocessing_" + param_type[5:6] + "__" + param_name[:-1]] = ev
+        elif param_type != "final_estimator_p":
+            #TODO regular express check ?
+            ev = safe_eval_es(literal)
+            preprocessors = [preprocessing.StandardScaler(), preprocessing.Binarizer(), preprocessing.Imputer(),
```
Comment (bgruening) [inline tools/sklearn/utils.py]: I find it more readable if you create a separate list for all the sub-groups to not do the slicing later preprocessors[0:8] and so on..

File: tools/sklearn/utils.py [hunk 12]
```diff
@@ -222,11 +254,112 @@ def get_search_params(params_builder):
...
+                            imblearn.over_sampling.SMOTENC(categorical_features=[], random_state=0, n_jobs=N_JOBS),
+                            imblearn.combine.SMOTEENN(random_state=0), imblearn.combine.SMOTETomek(random_state=0)]
+            i = 0
+            while i < len(ev):
+                obj = ev[i]
+                if obj is None:
+                    i = i + 1
+                    continue
+                elif obj == 'all':
+                    ev[i:i+1] = preprocessors
```
Comment (bgruening) [inline tools/sklearn/utils.py]: I guess I would collect the preprocessors in a separate list and do the merging and constructing of ev in the end. This is not really readable imho.

File: tools/sklearn/search_model_validation.xml (hunk 7 above)
Comment (qiagu) [reply tools/sklearn/search_model_validation.xml]: No better reason, just differ from other non-estimator eval. Do you have better idea?

File: tools/sklearn/search_model_validation.xml (hunk 9 above)
Comment (qiagu) [reply tools/sklearn/search_model_validation.xml]: The `-` is not part of the parameter, but cheetah markdown. The point in your second question is very good. If we do that, it will change the UI dramatically though. Let me think about it.

File: tools/sklearn/utils.py (hunk 10 above)
Comment (qiagu) [reply tools/sklearn/utils.py]: Oh, I was confused with `++` which is not supported in Python

File: tools/sklearn/utils.py (hunk 11 above)
Comment (qiagu) [reply tools/sklearn/utils.py]: Probably you're right. My solution may not be the best, or even one of those. I have no better idea on how to handle the hot numbers, and the mix at current moment. If you have one, a PR or commit is welcome

File: tools/sklearn/utils.py (hunk 12 above)
Comment (qiagu) [reply tools/sklearn/utils.py]: Oh, using a separate list by append/extent is a good idea.

Comment (qiagu) [reply tools/sklearn/utils.py]: Doing this `inplace` has a little memory benefit. :) I agree that making a newlist has more advantages.

**Comment by anuprulez** [comment]
That's a great work especially adding the `SMOTE` thing. I learnt about it in a different project sometime back to create additional data/oversampling using the similar distribution as the data itself. Thanks a lot @qiagu!

Do we need to recompute the results with this new version? It seems like a lot of change :)

**Comment by qiagu** [comment]
Thanks. I guess most of the dada don't need re-test. There could be float precision variance. I don't think it will change our figures much. For safe, we can run re-test of couple small datasets. Besides, we still can write the paper saying we tested in v19, IMHO

**Comment by qiagu** [comment]
I feel the `n_jobs =-1` helped the searchcv on EU sever before the parameter was overridden. I didn't have experience in my laptop, sorry

File: tools/sklearn/feature_selection.xml [hunk 13]
```diff
@@ -103,11 +103,14 @@ res.to_csv(path_or_buf="$outfile", sep='\t', index=False)
         <expand macro="feature_selection_all">
             <expand macro="fs_selectfrommodel_prefitted"/>
         </expand>
-        <expand macro="feature_selection_output_mothods" />
+        <param name="save" type="boolean" truevalue="booltrue" falsevalue="boolflase" checked="false" label="Save the fitted selector?"/>
```
Comment (anuprulez) [inline tools/sklearn/feature_selection.xml]: Again a typo: `falsevalue="boolflase"`. It should be `falsevalue="boolfalse"` :)

File: tools/sklearn/main_macros.xml [hunk 14]
```diff
@@ -824,14 +833,125 @@
...
+  <xml name="cv_reduced">
+    <conditional name="cv_selector">
+      <param name="selected_cv" type="select" label="Select the cv splitter:">
+        <expand macro="cv_splitter"/>
+      </param>
+      <expand macro="cv_splitter_options"/>
+    </conditional>
+  </xml>
+
+  <xml name="cv_n_splits" token_value="3" token_help="Number of folds. Muast be at least 2.">
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: Thanks @anuprulez

File: tools/sklearn/model_validation.xml [hunk 15]
```diff
@@ -67,7 +69,13 @@ y = read_columns(
 y=y.ravel()
 
 options = params["model_validation_functions"]["options"]
-options['cv'] = get_cv( options['cv'] )
+splitter, groups = get_cv(options.pop('cv_selector'))
+if groups is None:
+    options['cv'] = splitter
+elif groups == "":
+    options['cv'] = list( splitter.split(X, y, groups=None) )
```
Comment (anuprulez) [inline tools/sklearn/model_validation.xml]: I see two empty spaces before and after `list`. But, in line `72` the spaces are not present. In general, we should stick to a similar standard.

**Comment by qiagu** [comment]
I plan to remove the cache. It only helps small number of algorithms and there could be alternatives. What do you think? @bgruening @anuprulez

File: tools/sklearn/discriminant.xml [hunk 16]
```diff
@@ -106,7 +106,7 @@ with open("$outfile_fit", 'wb') as out_handler:
             <param name="selected_algorithm" value="LinearDiscriminantAnalysis"/>
             <param name="solver" value="svd" />
             <param name="store_covariance" value="True"/>
-            <output name="outfile_fit" file="lda_model01" compare="sim_size" delta="500"/>
+            <output name="outfile_fit" file="lda_model01" compare="sim_size" delta="1"/>
```
Comment (anuprulez) [inline tools/sklearn/discriminant.xml]: The value of `delta` varies throughout. Some places it is `1`, other places `5`. Is it possible to use just one value?

**Comment by anuprulez** [comment]
@qiagu @bgruening 

It is a great work. Thanks!

I think more and more people use these tools in workflows and the workflows already save the parameter values (reusable). Moreover, for standalone tool usage as well a user can rerun the job with the same parameters as the original job. The caching of these parameter values would be superfluous in my opinion if I understand the purpose of caching correctly.

**Comment by qiagu** [comment]
The caching could increase the speed of RFE when tuning the `n_features_to_select`, but we can use RFECV which is more efficient.

File: tools/sklearn/discriminant.xml (hunk 16 above)
Comment (qiagu) [reply tools/sklearn/discriminant.xml]: The file size varies a lot. For MB sized file, dalta 500 is good enough, but for a file of hundreds of Byte, delta 1 gives better accuracy.

**Comment by bgruening** [comment]
@qiagu I'm with you here. It was good to explore the caching but this can wait for later and we can discuss this in more detail and maybe can come up with some infrastructure in Galaxy as well. Ideally, this should be not exposed to the user but handled in a smart way by the system.

File: tools/sklearn/discriminant.xml (hunk 16 above)
Comment (anuprulez) [reply tools/sklearn/discriminant.xml]: I am not sure I understood it completely. Here `sim_size` is being used under tests. How it is related to the performance of an algorithm?

Comment (qiagu) [reply tools/sklearn/discriminant.xml]: Sorry. The accuracy doesn't relate to performance, but points to the right output. The delta, in my understanding, is the difference between the output size and the test data file size. If the size of output is expected to be in hundreds of Bytes, delta 500 may provides little assurance that the output from the test is the same as the test data file.

File: tools/sklearn/pipeline.xml [hunk 17]
```diff
@@ -80,26 +92,52 @@ def get_component(input_json, check_none=False):
...
+            has_imblearn = True
 
 # Set up final estimator and add to pipeline.
 estimator_json = params["final_estimator"]['estimator_selector']
-estimator = get_estimator(estimator_json)
-
-pipeline_steps.append( ('estimator', estimator) )
+if estimator_json['selected_module'] == 'none':
+    if len(pipeline_steps) == 0:
+        sys.exit("No pipeline steps!")
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/pipeline.xml]: Thanks @bgruening

**Comment by bgruening** [comment]
@qiagu I will try to release a new version of xgboost, but this can take some time. We can only increase the version number of dependencies if there is a conda package available. For xgboost 0.81 there is none at the moment. There are also a bunch of API changes as documented here: https://github.com/dmlc/xgboost/releases

**Comment by bgruening** [comment]
Ok, I was able to create this conda-package ... will restart in a minute the tests.

**Comment by qiagu** [comment]
Looks there are a lot of hassles. Thank you @bgruening

**Comment by bgruening** [comment]
We could also go back to the older xgboost version, I think.

**Comment by bgruening** [comment]
Thanks for reverting. Let's get this in.

@qiagu this is outstanding work, thanks a lot!

**Comment by bgruening** [comment]
@qiagu should be updated :)
